File: app/graph/orchestrator.py
"""Outer LangGraph: session-aware orchestrator with signal monitoring."""

import logging
import os
from datetime import datetime, timedelta, timezone

from langgraph.graph import StateGraph

from app.db import init_db, get_latest_snapshot, get_snapshots
from app.fetch_portfolio import load_portfolio
from app.graph.analysis import build_analysis_graph, decide_depth
from app.graph.session import detect_session
from app.graph.signals import detect_signals_for_ticker
from app.graph.state import (
    AnalysisState,
    OrchestratorState,
    Session,
    SignalTier,
    TickerSignal,
)

log = logging.getLogger(__name__)

COOLDOWN_MIN = int(os.getenv("SIGNAL_COOLDOWN_MIN", "15"))
SCHEDULED_INTERVAL_MIN = 30


def route_decision(
    session: Session,
    signals: list[TickerSignal],
    pending_batch: list[str],
    scheduled: bool,
) -> str:
    """Decide what to do: skip, batch, or immediate."""
    if session == Session.CLOSED:
        return "skip"

    has_major = any(s["tier"] == SignalTier.MAJOR for s in signals)
    if has_major:
        return "immediate"

    has_minor = any(s["tier"] == SignalTier.MINOR for s in signals)
    if has_minor or scheduled or pending_batch:
        return "batch"

    return "skip"


def get_sleep_interval(session: Session) -> int:
    """Return sleep interval in seconds based on session."""
    active_interval = int(os.getenv("GRAPH_ACTIVE_INTERVAL", "300"))
    idle_interval = int(os.getenv("GRAPH_IDLE_INTERVAL", "1800"))

    if session in (Session.SESSION_1, Session.SESSION_2):
        return active_interval
    if session in (Session.LUNCH, Session.PRE_MARKET):
        return 900
    return idle_interval


def should_run_scheduled(last_scheduled: datetime, interval_min: int = SCHEDULED_INTERVAL_MIN) -> bool:
    """Check if enough time has passed for a scheduled batch run."""
    elapsed = (datetime.now(timezone.utc) - last_scheduled).total_seconds() / 60
    return elapsed >= interval_min


def is_on_cooldown(ticker: str, last_run: dict[str, datetime], cooldown_min: int = COOLDOWN_MIN) -> bool:
    """Check if a ticker is on cooldown."""
    last = last_run.get(ticker)
    if last is None:
        return False
    elapsed = (datetime.now(timezone.utc) - last).total_seconds() / 60
    return elapsed < cooldown_min


def _node_detect_session(state: OrchestratorState) -> dict:
    """Detect current market session."""
    session = detect_session()
    return {"current_session": session, "last_check": datetime.now(timezone.utc)}


def _node_check_signals(state: OrchestratorState) -> dict:
    """Check all active tickers for signals."""
    session = state["current_session"]
    if session == Session.CLOSED:
        return {"signals": []}

    portfolio = load_portfolio()
    all_signals: list[TickerSignal] = []

    for ticker in portfolio:
        if is_on_cooldown(ticker, state.get("last_run", {})):
            continue
        snapshot = get_latest_snapshot(ticker)
        if not snapshot:
            continue

        prev_snapshots = get_snapshots(ticker, limit=2)
        prev = prev_snapshots[1] if len(prev_snapshots) > 1 else None

        ticker_signals = detect_signals_for_ticker(snapshot, prev)
        all_signals.extend(ticker_signals)

    return {"signals": all_signals}


def _node_route(state: OrchestratorState) -> dict:
    """Route decision — sets _route key for conditional edges."""
    session = state["current_session"]
    signals = state.get("signals", [])
    pending = state.get("pending_batch", [])
    scheduled = should_run_scheduled(
        state.get("last_check", datetime.min.replace(tzinfo=timezone.utc)),
    )

    decision = route_decision(session, signals, pending, scheduled)

    if decision == "batch":
        minor_tickers = [s["ticker"] for s in signals if s["tier"] == SignalTier.MINOR]
        batch_tickers = list(set(pending + minor_tickers))
        if scheduled and not batch_tickers:
            portfolio = load_portfolio()
            batch_tickers = list(portfolio.keys())
        return {"_route": decision, "pending_batch": batch_tickers}

    if decision == "immediate":
        major_tickers = [s["ticker"] for s in signals if s["tier"] == SignalTier.MAJOR]
        return {"_route": decision, "pending_batch": major_tickers}

    return {"_route": decision}


def _node_run_analysis(state: OrchestratorState) -> dict:
    """Run analysis pipeline on pending tickers."""
    tickers = state.get("pending_batch", [])
    if not tickers:
        return {"pending_batch": []}

    session = state["current_session"]
    overall_tier = SignalTier.MAJOR if any(
        s["tier"] == SignalTier.MAJOR for s in state.get("signals", [])
    ) else SignalTier.MINOR if state.get("signals") else None

    depth = decide_depth(session, overall_tier)

    analysis_state = AnalysisState(
        tickers=tickers,
        depth=depth,
        session=session,
        results={},
        errors={},
    )

    analysis_graph = build_analysis_graph()
    analysis_graph.invoke(analysis_state)

    now = datetime.now(timezone.utc)
    last_run = dict(state.get("last_run", {}))
    for t in tickers:
        last_run[t] = now

    return {"pending_batch": [], "signals": [], "last_run": last_run}


def _node_skip(state: OrchestratorState) -> dict:
    """No-op node for skip route."""
    return {}


def _route_edge(state: OrchestratorState) -> str:
    """Conditional edge function for routing."""
    return state.get("_route", "skip")


def build_orchestrator_graph():
    """Build the outer orchestrator graph."""
    graph = StateGraph(OrchestratorState)

    graph.add_node("detect_session", _node_detect_session)
    graph.add_node("check_signals", _node_check_signals)
    graph.add_node("route", _node_route)
    graph.add_node("run_analysis", _node_run_analysis)
    graph.add_node("skip", _node_skip)

    graph.add_edge("__start__", "detect_session")
    graph.add_edge("detect_session", "check_signals")
    graph.add_edge("check_signals", "route")
    graph.add_conditional_edges(
        "route",
        _route_edge,
        {
            "immediate": "run_analysis",
            "batch": "run_analysis",
            "skip": "skip",
        },
    )
    graph.add_edge("run_analysis", "__end__")
    graph.add_edge("skip", "__end__")

    return graph.compile()

File: app/src/graph/orchestrator.ts
// app/src/graph/orchestrator.ts
import { StateGraph, START, END, Annotation } from '@langchain/langgraph'
import type { OrchestratorState, Session, TickerSignal } from './state'
import { detectSession } from './session'
import { detectSignalsForTicker, filterCooledSignals } from './signals'
import { loadPortfolio, getWatchlist, getLatestSnapshots, getSnapshotBefore } from '../db/db'
import { runPriceRefresh } from '../services/portfolio'
import { buildAnalysisGraph, decideDepth } from './analysis'

const OrchestratorAnnotation = Annotation.Root({
  current_session: Annotation<Session>,
  last_session:    Annotation<Session | null>,
  last_check:      Annotation<string>,
  last_scheduled:  Annotation<string | null>,
  signals:         Annotation<TickerSignal[]>,
  signal_cooldowns: Annotation<Record<string, string>>,
  pending_batch:   Annotation<string[]>,
  last_run:        Annotation<string | null>,
  last_news_fetch: Annotation<string | null>,
  _route:          Annotation<string | undefined>,
})

type OrchestratorAnnotationType = typeof OrchestratorAnnotation.State

async function detectSessionNode(
  state: OrchestratorAnnotationType,
): Promise<Partial<OrchestratorAnnotationType>> {
  const session = detectSession()
  return { current_session: session, last_session: state.current_session, last_check: new Date().toISOString() }
}

async function refreshPricesNode(
  _state: OrchestratorAnnotationType,
): Promise<Partial<OrchestratorAnnotationType>> {
  await runPriceRefresh()
  return {}
}

async function fetchNewsNode(
  _state: OrchestratorAnnotationType,
): Promise<Partial<OrchestratorAnnotationType>> {
  return { last_news_fetch: new Date().toISOString() }
}

async function checkSignalsNode(
  state: OrchestratorAnnotationType,
): Promise<Partial<OrchestratorAnnotationType>> {
  const portfolio = await loadPortfolio()
  const wl = await getWatchlist()
  const tickers = [...new Set([...Object.keys(portfolio), ...wl.map(w => w.ticker)])]
  const weekAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000)

  // One batched query for all latest snapshots, then fan out the per-ticker
  // "week ago" lookups concurrently — avoids the 2N serial round-trips this
  // node used to make every runner cycle.
  const latest = new Map((await getLatestSnapshots()).map(s => [s.ticker, s]))
  const withSnap = tickers.filter(t => latest.has(t))
  const prevSnaps = await Promise.all(withSnap.map(t => getSnapshotBefore(t, weekAgo)))

  const signals: TickerSignal[] = []
  withSnap.forEach((ticker, i) => {
    signals.push(...detectSignalsForTicker(ticker, latest.get(ticker)!, prevSnaps[i]))
  })

  return { signals }
}

function routeNode(
  state: OrchestratorAnnotationType,
): Partial<OrchestratorAnnotationType> {
  const { current_session, last_session, signals } = state

  if (current_session === 'CLOSED') {
    return { _route: 'skip' }
  }

  // Only MAJOR signals outside their cooldown window trigger a signal run;
  // cooled signals fall through so the scheduled cadence isn't starved by a
  // stock pinned at a big day move (e.g. ARA) all session.
  const majorSignals = filterCooledSignals(
    signals.filter(s => s.tier === 'MAJOR'),
    state.signal_cooldowns ?? {},
  )
  if (majorSignals.length > 0) {
    const stampedAt = new Date().toISOString()
    const signal_cooldowns = { ...state.signal_cooldowns }
    for (const s of majorSignals) signal_cooldowns[s.ticker] = stampedAt
    return { _route: 'run_analysis', pending_batch: majorSignals.map(s => s.ticker), signal_cooldowns }
  }

  // Fire once at every session boundary (PRE_MARKET→SESSION_1, SESSION_1→LUNCH, etc.)
  if (last_session !== null && last_session !== current_session) {
    console.log(`[orchestrator] session transition ${last_session} → ${current_session}, triggering analysis`)
    return { _route: 'run_analysis', pending_batch: [], last_scheduled: new Date().toISOString() }
  }

  // Check if scheduled run is due (every GRAPH_ANALYSIS_INTERVAL minutes).
  // Deliberately separate from GRAPH_ACTIVE_INTERVAL (runner loop sleep):
  // the loop polls prices/signals often, analysis runs on its own cadence.
  const intervalMin = Number(process.env.GRAPH_ANALYSIS_INTERVAL ?? 30)
  const lastSched = state.last_scheduled ? new Date(state.last_scheduled) : null
  const elapsed = lastSched ? (Date.now() - lastSched.getTime()) / 60_000 : Infinity

  if (elapsed >= intervalMin) {
    return { _route: 'run_analysis', pending_batch: [], last_scheduled: new Date().toISOString() }
  }

  return { _route: 'skip' }
}

async function runAnalysisNode(
  state: OrchestratorAnnotationType,
): Promise<Partial<OrchestratorAnnotationType>> {
  // pending_batch is non-empty only on the signal route; a cooled MAJOR signal
  // may still sit in state.signals during a scheduled run, so batch decides tier.
  const isSpike = state.pending_batch.length > 0
  const depth = decideDepth(state.current_session, isSpike ? 'MAJOR' : 'MINOR')

  // Signal-triggered tickers go to the multi-agent deep-run queue when enabled
  // (drained by graph/worker.ts); anything that fails to enqueue falls back to
  // the inline single-pass so a queue outage never silences a MAJOR signal.
  let inlineBatch = state.pending_batch
  if (isSpike && process.env.DEEP_RUNS_ENABLED === 'true') {
    const { enqueueDeepRun } = await import('../services/deepRun.js')
    const fallback: string[] = []
    for (const ticker of state.pending_batch) {
      try {
        await enqueueDeepRun(ticker)
      } catch (err) {
        console.error(`[orchestrator] deep-run enqueue failed for ${ticker}, falling back inline:`, err)
        fallback.push(ticker)
      }
    }
    if (fallback.length === 0) {
      return { last_run: new Date().toISOString(), signals: [] }
    }
    inlineBatch = fallback
  }

  const graph = buildAnalysisGraph()
  await graph.invoke({
    tickers: inlineBatch,
    depth,
    session: state.current_session,
    // Telegram alerts are spike-only; scheduled/session-boundary runs stay silent
    alerts: isSpike ? 'spike' as const : 'silent' as const,
    results: {},
    errors: {},
  })

  return { last_run: new Date().toISOString(), signals: [] }
}

async function skipNode(
  _state: OrchestratorAnnotationType,
): Promise<Partial<OrchestratorAnnotationType>> {
  return {}
}

function routeDecision(state: OrchestratorAnnotationType): string {
  return state._route ?? 'skip'
}

export function buildOrchestratorGraph() {
  const graph = new StateGraph(OrchestratorAnnotation)
    .addNode('detect_session',  detectSessionNode)
    .addNode('refresh_prices',  refreshPricesNode)
    .addNode('fetch_news',      fetchNewsNode)
    .addNode('check_signals',   checkSignalsNode)
    .addNode('route',           (s: OrchestratorAnnotationType) => ({ ...routeNode(s) }))
    .addNode('run_analysis',    runAnalysisNode)
    .addNode('skip',            skipNode)
    .addEdge(START, 'detect_session')
    .addEdge('detect_session', 'refresh_prices')
    .addEdge('refresh_prices', 'fetch_news')
    .addEdge('fetch_news', 'check_signals')
    .addEdge('check_signals', 'route')
    .addConditionalEdges('route', routeDecision, {
      run_analysis: 'run_analysis',
      skip: 'skip',
    })
    .addEdge('run_analysis', END)
    .addEdge('skip', END)

  return graph.compile()
}

// Re-export for runner.ts compatibility
export type { OrchestratorState }
